radio_beam/commonbeam.py

```diff
@@ -0,0 +1,522 @@
+        maj_step = (maj_range[1] - maj_range[0]) / brute_steps
+        min_range = [beams.minor.min(), 1.5 * beams.major.max()]
+        min_step = (min_range[1] - min_range[0]) / brute_steps
+        rranges = (slice(maj_range[0], maj_range[1], maj_step),
+                   slice(min_range[0], min_range[1], min_step),
+                   slice(0, 179.9, 180. / brute_steps))
+        result = opt.brute(myobjective_regularized, rranges,
+                           args=(bmaj, bmin, bpa),
+                           full_output=True,
+                           finish=opt.fmin)
+        params = result[0]
+
+    else:
+        result = opt.minimize(myobjective_regularized, p0,
+                              method='Nelder-Mead',
+                              args=(bmaj, bmin, bpa),
+                              options=optdict,
+                              tol=1e-14)
+        params = result.x
+
+        if verbose:
+            print(result.viewitems())
+
+        if not result.success:
+            raise Warning("Optimization failed")
+
+    com_beam = Beam(params[0] * beams.major.unit,
+                    params[1] * beams.major.unit,
+                    (params[2] % np.pi) * u.rad)
+
+    # Test if it deconvolves all
+    if not fits_in_largest(beams, com_beam):
+        raise BeamError("Could not find common beam to deconvolve all beams.")
+
+    return com_beam
+
+
+def fits_in_largest(beams, large_beam=None):
+    '''
+    Test if all beams can be deconvolved by the largest beam
+    '''
+
+    if large_beam is None:
+        large_beam = beams.largest_beam()
+
+    for beam in beams:
+        if large_beam == beam:
+            continue
+        deconv_beam = large_beam.deconvolve(beam,
+                                            failure_returns_pointlike=True)
+
+        if not deconv_beam.isfinite:
+            return False
+
+    return True
+
+
+def getMinVolEllipse(P, tolerance=1e-5, maxiter=1e5):
+    """
+    Use the Khachiyan Algorithm to compute that minimum volume ellipsoid.
+
+    Adapted code from: https://github.com/minillinim/ellipsoid/blob/master/ellipsoid.py
+
+    This code was further adapted from:
+
+    Based on work by Nima Moshtagh
+    http://www.mathworks.com/matlabcentral/fileexchange/9542
+    and also by looking at:
+    http://cctbx.sourceforge.net/current/python/scitbx.math.minimum_covering_ellipsoid.html
+    Which is based on the first reference anyway!
+
+    Parameters
+    ----------
+    P : `~numpy.ndarray`
+        Points to compute solution.
+    tolerance : float, optional
+        Allowed error range in the Khachiyan Algorithm. Decreasing the
+        tolerance by an order of magnitude requires an order of magnitude
+        more iterations to converge.
+    maxiter : int, optional
+        Maximum iterations.
+
+    Returns
+    -------
+    center : `~numpy.ndarray`
+        Center point of the ellipse. Is required to be smaller than the
+        tolerance.
```

Comment:
So this algorithm is more general than our case, in which the beams are forced to have a common center?

Reply:
Yes. I tried to re-express it without needing to solve for the centre, but I couldn't get it right. I added a check that requires the center be within the error tolerance of (0, 0). It's been much smaller than the tolerance (~1e-10 in x and y) in all of my tests so I don't expect this to be an issue.